akka-remote/src/main/scala/akka/remote/Remoting.scala

```diff
@@ -548,18 +555,28 @@ private[remote] class EndpointManager(conf: Config, log: LoggingAdapter) extends
             "address cannot be quarantined without knowing the UID, gating instead for {} ms.",
             address, settings.RetryGateClosedFor.toMillis)
           endpoints.markAsFailed(endpoint, Deadline.now + settings.RetryGateClosedFor)
-        case (Some(Pass(endpoint, Some(currentUid), _)), Some(quarantineUid)) if currentUid == quarantineUid ⇒
-          context.stop(endpoint)
-        case _ ⇒
-        // Do nothing, because either:
-        // A: we don't know yet the UID of the writer, it will be checked against current quarantine state later
-        // B: we know the UID, but it does not match with the UID to be quarantined
+        case (Some(Pass(endpoint, uidOption, refuseUidOption)), Some(quarantineUid)) ⇒
+          uidOption match {
+            case Some(uid) if uid == quarantineUid ⇒
+              endpoints.markAsQuarantined(address, quarantineUid, Deadline.now + settings.QuarantineDuration)
+              eventPublisher.notifyListeners(QuarantinedEvent(address, quarantineUid))
+              context.stop(endpoint)
+            // or it does not match with the UID to be quarantined
+            case None if !refuseUidOption.contains(quarantineUid) ⇒
+              // the quarantine uid may be got fresh by cluster gossip, so update refuseUid for late handle when the writer got uid
+              endpoints.registerWritableEndpointRefuseUid(address, quarantineUid)
+            case _ ⇒ //the quarantine uid has lost the race with some failure, do nothing
+          }
+        case (_, Some(quarantinedUid)) ⇒
+          // the current state is gated or quarantined, and we know the UID, update
+          endpoints.markAsQuarantined(address, quarantinedUid, Deadline.now + settings.QuarantineDuration)
+          eventPublisher.notifyListeners(QuarantinedEvent(address, quarantinedUid))
```

Comment:
what we can only do is avoid to publish the same quarantinedUid event twice. the above code keep the same logic as the orginal one( the same "problem" exists), but i didn't know whether anybody report it before.

I prefer to add the recheck logic, because it happens when the cluster watcher used( frequently used by us).
e.g. rds throws Hopeless -> manager make node quarantined & publish event -> cluster down and remove the node -> cluster watcher quarantine the node -> manager publish again